// @ellipsis-labs/phoenix-sdk/psm-src/instructions/ChangeMarketStatus.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from "@metaplex-foundation/beet";
import * as web3 from "@solana/web3.js";
import { MarketStatus, marketStatusBeet } from "../../src/types/MarketStatus";

/**
 * @category Instructions
 * @category ChangeMarketStatus
 * @category generated
 */
export type ChangeMarketStatusInstructionArgs = {
  marketStatus: MarketStatus;
};
/**
 * @category Instructions
 * @category ChangeMarketStatus
 * @category generated
 */
export const ChangeMarketStatusStruct = new beet.BeetArgsStruct<
  ChangeMarketStatusInstructionArgs & {
    instructionDiscriminator: number;
  }
>(
  [
    ["instructionDiscriminator", beet.u8],
    ["marketStatus", marketStatusBeet],
  ],
  "ChangeMarketStatusInstructionArgs"
);
/**
 * Accounts required by the _ChangeMarketStatus_ instruction
 *
 * @property [] phoenixProgram Phoenix program
 * @property [] logAuthority Phoenix log authority
 * @property [_writable_] market This account holds the market state
 * @property [_writable_, **signer**] marketAuthority The market_authority account must sign to change market status
 * @property [**signer**] seatManagerAuthority The seat manager account must sign change market status
 * @category Instructions
 * @category ChangeMarketStatus
 * @category generated
 */
export type ChangeMarketStatusInstructionAccounts = {
  phoenixProgram: web3.PublicKey;
  logAuthority: web3.PublicKey;
  market: web3.PublicKey;
  marketAuthority: web3.PublicKey;
  seatManagerAuthority: web3.PublicKey;
};

export const changeMarketStatusInstructionDiscriminator = 8;

/**
 * Creates a _ChangeMarketStatus_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category ChangeMarketStatus
 * @category generated
 */
export function createChangeMarketStatusInstruction(
  accounts: ChangeMarketStatusInstructionAccounts,
  args: ChangeMarketStatusInstructionArgs,
  programId = new web3.PublicKey("PSMxQbAoDWDbvd9ezQJgARyq6R9L5kJAasaLDVcZwf1")
) {
  const [data] = ChangeMarketStatusStruct.serialize({
    instructionDiscriminator: changeMarketStatusInstructionDiscriminator,
    ...args,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.phoenixProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.logAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.market,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.marketAuthority,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.seatManagerAuthority,
      isWritable: false,
      isSigner: true,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  });
  return ix;
}
