// @ellipsis-labs/phoenix-sdk/psm-src/instructions/ConfirmRenounceSeatManagerAuthority.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from "@metaplex-foundation/beet";
import * as web3 from "@solana/web3.js";

/**
 * @category Instructions
 * @category ConfirmRenounceSeatManagerAuthority
 * @category generated
 */
export const ConfirmRenounceSeatManagerAuthorityStruct =
  new beet.BeetArgsStruct<{ instructionDiscriminator: number }>(
    [["instructionDiscriminator", beet.u8]],
    "ConfirmRenounceSeatManagerAuthorityInstructionArgs"
  );
/**
 * Accounts required by the _ConfirmRenounceSeatManagerAuthority_ instruction
 *
 * @property [_writable_] seatManager This account holds the seat manager state
 * @property [**signer**] seatManagerAuthority The seat manager authority must sign to renounce the seat manager authority
 * @category Instructions
 * @category ConfirmRenounceSeatManagerAuthority
 * @category generated
 */
export type ConfirmRenounceSeatManagerAuthorityInstructionAccounts = {
  seatManager: web3.PublicKey;
  seatManagerAuthority: web3.PublicKey;
};

export const confirmRenounceSeatManagerAuthorityInstructionDiscriminator = 11;

/**
 * Creates a _ConfirmRenounceSeatManagerAuthority_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category ConfirmRenounceSeatManagerAuthority
 * @category generated
 */
export function createConfirmRenounceSeatManagerAuthorityInstruction(
  accounts: ConfirmRenounceSeatManagerAuthorityInstructionAccounts,
  programId = new web3.PublicKey("PSMxQbAoDWDbvd9ezQJgARyq6R9L5kJAasaLDVcZwf1")
) {
  const [data] = ConfirmRenounceSeatManagerAuthorityStruct.serialize({
    instructionDiscriminator:
      confirmRenounceSeatManagerAuthorityInstructionDiscriminator,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.seatManager,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.seatManagerAuthority,
      isWritable: false,
      isSigner: true,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  });
  return ix;
}
