// @ellipsis-labs/phoenix-sdk/psm-src/instructions/EvictSeat.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as splToken from "@solana/spl-token";
import * as beet from "@metaplex-foundation/beet";
import * as web3 from "@solana/web3.js";

/**
 * @category Instructions
 * @category EvictSeat
 * @category generated
 */
export const EvictSeatStruct = new beet.BeetArgsStruct<{
  instructionDiscriminator: number;
}>([["instructionDiscriminator", beet.u8]], "EvictSeatInstructionArgs");
/**
 * Accounts required by the _EvictSeat_ instruction
 *
 * @property [] phoenixProgram Phoenix program
 * @property [] logAuthority Phoenix log authority
 * @property [_writable_] market This account holds the market state
 * @property [_writable_] seatManager The seat manager account must sign to evict a seat
 * @property [_writable_] seatDepositCollector Collects deposits for claiming new seats and refunds for evicting seats
 * @property [] baseMint
 * @property [] quoteMint
 * @property [_writable_] baseVault
 * @property [_writable_] quoteVault
 * @property [] associatedTokenAccountProgram Associated token account program
 * @property [**signer**] signer
 * @property [_writable_] trader
 * @property [] seat The trader's PDA seat account, seeds are [b'seat', market_address, trader_address]
 * @property [_writable_] baseAccount The trader's associated token account for the base mint
 * @property [_writable_] quoteAccount The trader's associated token account for the quote mint
 * @property [_writable_] baseAccountBackup Non-ATA token account for the base mint, in case the ATA owner is no longer the trader
 * @property [_writable_] quoteAccountBackup Non-ATA token account for the quote mint, in case the ATA owner is no longer the trader
 * @category Instructions
 * @category EvictSeat
 * @category generated
 */
export type EvictSeatInstructionAccounts = {
  phoenixProgram: web3.PublicKey;
  logAuthority: web3.PublicKey;
  market: web3.PublicKey;
  seatManager: web3.PublicKey;
  seatDepositCollector: web3.PublicKey;
  baseMint: web3.PublicKey;
  quoteMint: web3.PublicKey;
  baseVault: web3.PublicKey;
  quoteVault: web3.PublicKey;
  associatedTokenAccountProgram: web3.PublicKey;
  tokenProgram?: web3.PublicKey;
  systemProgram?: web3.PublicKey;
  signer: web3.PublicKey;
  trader: web3.PublicKey;
  seat: web3.PublicKey;
  baseAccount: web3.PublicKey;
  quoteAccount: web3.PublicKey;
  baseAccountBackup: web3.PublicKey;
  quoteAccountBackup: web3.PublicKey;
};

export const evictSeatInstructionDiscriminator = 3;

/**
 * Creates a _EvictSeat_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category EvictSeat
 * @category generated
 */
export function createEvictSeatInstruction(
  accounts: EvictSeatInstructionAccounts,
  programId = new web3.PublicKey("PSMxQbAoDWDbvd9ezQJgARyq6R9L5kJAasaLDVcZwf1")
) {
  const [data] = EvictSeatStruct.serialize({
    instructionDiscriminator: evictSeatInstructionDiscriminator,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.phoenixProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.logAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.market,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.seatManager,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.seatDepositCollector,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.baseMint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.quoteMint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.baseVault,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.quoteVault,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.associatedTokenAccountProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.tokenProgram ?? splToken.TOKEN_PROGRAM_ID,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.trader,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.seat,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.baseAccount,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.quoteAccount,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.baseAccountBackup,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.quoteAccountBackup,
      isWritable: true,
      isSigner: false,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  });
  return ix;
}
