// @ellipsis-labs/phoenix-sdk/psm-src/instructions/NameSuccessor.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from "@metaplex-foundation/beet";
import * as web3 from "@solana/web3.js";

/**
 * @category Instructions
 * @category NameSuccessor
 * @category generated
 */
export const NameSuccessorStruct = new beet.BeetArgsStruct<{
  instructionDiscriminator: number;
}>([["instructionDiscriminator", beet.u8]], "NameSuccessorInstructionArgs");
/**
 * Accounts required by the _NameSuccessor_ instruction
 *
 * @property [_writable_] seatManager This account holds the seat manager state
 * @property [**signer**] seatManagerAuthority The seat manager account must sign name a successor
 * @property [] successor The new authority account
 * @category Instructions
 * @category NameSuccessor
 * @category generated
 */
export type NameSuccessorInstructionAccounts = {
  seatManager: web3.PublicKey;
  seatManagerAuthority: web3.PublicKey;
  successor: web3.PublicKey;
};

export const nameSuccessorInstructionDiscriminator = 6;

/**
 * Creates a _NameSuccessor_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category NameSuccessor
 * @category generated
 */
export function createNameSuccessorInstruction(
  accounts: NameSuccessorInstructionAccounts,
  programId = new web3.PublicKey("PSMxQbAoDWDbvd9ezQJgARyq6R9L5kJAasaLDVcZwf1")
) {
  const [data] = NameSuccessorStruct.serialize({
    instructionDiscriminator: nameSuccessorInstructionDiscriminator,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.seatManager,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.seatManagerAuthority,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.successor,
      isWritable: false,
      isSigner: false,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  });
  return ix;
}
