// @ellipsis-labs/phoenix-sdk/psm-src/instructions/RemoveDesignatedMarketMaker.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from "@metaplex-foundation/beet";
import * as web3 from "@solana/web3.js";

/**
 * @category Instructions
 * @category RemoveDesignatedMarketMaker
 * @category generated
 */
export const RemoveDesignatedMarketMakerStruct = new beet.BeetArgsStruct<{
  instructionDiscriminator: number;
}>(
  [["instructionDiscriminator", beet.u8]],
  "RemoveDesignatedMarketMakerInstructionArgs"
);
/**
 * Accounts required by the _RemoveDesignatedMarketMaker_ instruction
 *
 * @property [] market This account holds the market state
 * @property [_writable_] seatManager This account holds the seat manager state
 * @property [] trader
 * @property [**signer**] seatManagerAuthority The seat manager authority account must sign to remove a DMM
 * @category Instructions
 * @category RemoveDesignatedMarketMaker
 * @category generated
 */
export type RemoveDesignatedMarketMakerInstructionAccounts = {
  market: web3.PublicKey;
  seatManager: web3.PublicKey;
  trader: web3.PublicKey;
  seatManagerAuthority: web3.PublicKey;
};

export const removeDesignatedMarketMakerInstructionDiscriminator = 5;

/**
 * Creates a _RemoveDesignatedMarketMaker_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category RemoveDesignatedMarketMaker
 * @category generated
 */
export function createRemoveDesignatedMarketMakerInstruction(
  accounts: RemoveDesignatedMarketMakerInstructionAccounts,
  programId = new web3.PublicKey("PSMxQbAoDWDbvd9ezQJgARyq6R9L5kJAasaLDVcZwf1")
) {
  const [data] = RemoveDesignatedMarketMakerStruct.serialize({
    instructionDiscriminator:
      removeDesignatedMarketMakerInstructionDiscriminator,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.market,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.seatManager,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.trader,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.seatManagerAuthority,
      isWritable: false,
      isSigner: true,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  });
  return ix;
}
